playback/src/local_file.rs

```diff
@@ -0,0 +1,182 @@
+use librespot_core::{Error, SpotifyUri};
+use regex::{Captures, Regex};
+use std::sync::LazyLock;
+use std::{
+    collections::HashMap,
+    fs,
+    fs::File,
+    io,
+    path::{Path, PathBuf},
+    time::Duration,
+};
+use symphonia::{
+    core::formats::FormatOptions,
+    core::io::MediaSourceStream,
+    core::meta::{MetadataOptions, StandardTagKey, Tag},
+    core::probe::{Hint, ProbeResult},
+};
+
+// "Spotify supports .mp3, .mp4, and .m4p files. It doesn’t support .mp4 files that contain video,
+// or the iTunes lossless format (M4A)."
+// https://community.spotify.com/t5/FAQs/Local-Files/ta-p/5186118
+//
+// There are some indications online that FLAC is supported, so check for this as well.
+const SUPPORTED_FILE_EXTENSIONS: &[&str; 4] = &["mp3", "mp4", "m4p", "flac"];
+
+#[derive(Default)]
+pub struct LocalFileLookup(HashMap<SpotifyUri, PathBuf>);
+
+impl LocalFileLookup {
+    pub fn get(&self, uri: &SpotifyUri) -> Option<&Path> {
+        self.0.get(uri).map(|p| p.as_path())
+    }
+}
+
+pub fn create_local_file_lookup(directories: &[PathBuf]) -> LocalFileLookup {
+    let mut lookup = LocalFileLookup(HashMap::new());
+
+    for path in directories {
+        if !path.is_dir() {
+            warn!(
+                "Ignoring local file source {}: not a directory",
+                path.display()
+            );
+            continue;
+        }
+
+        if let Err(e) = visit_dir(path, &mut lookup) {
+            warn!(
+                "Failed to load entries from local file source {}: {}",
+                path.display(),
+                e
+            );
+        }
+    }
+
+    lookup
+}
+
+fn visit_dir(dir: &Path, accumulator: &mut LocalFileLookup) -> io::Result<()> {
+    for entry in fs::read_dir(dir)? {
+        let path = entry?.path();
+        if path.is_dir() {
+            visit_dir(&path, accumulator)?;
+        } else {
+            let Some(extension) = path.extension().and_then(|e| e.to_str()) else {
+                continue;
+            };
+
+            let lowercase_extension = extension.to_lowercase();
+
+            if SUPPORTED_FILE_EXTENSIONS.contains(&lowercase_extension.as_str()) {
+                let uri = match get_uri_from_file(path.as_path(), extension) {
+                    Ok(uri) => uri,
+                    Err(e) => {
+                        warn!(
+                            "Failed to determine URI of local file {}: {}",
+                            path.display(),
+                            e
+                        );
+                        continue;
+                    }
+                };
+
+                accumulator.0.insert(uri, path);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn get_uri_from_file(audio_path: &Path, extension: &str) -> Result<SpotifyUri, Error> {
```

Comment:
`new_with_probe` has a similar entry, would it make sense to reduce some of the logic into a method that just returns the `probed` value? And maybe changing `extension` to `file_extension` so that it's also clear from the outside what is expected?

Reply:
Na, if those parts are different now, just let them be separated. Funny enough, I think we will resolved #1487 with this PR then?